test/integration/client/exec_test.go

```diff
@@ -407,6 +382,135 @@ func TestExecPlugin(t *testing.T) {
+	updates []oldNew
+	deletes []interface{}
+}
+
+func (es *informerSpy) OnAdd(obj interface{}) { es.adds = append(es.adds, obj) }
+
+func (es *informerSpy) OnUpdate(old, new interface{}) {
+	es.updates = append(es.updates, oldNew{old: old, new: new})
+}
+
+func (es *informerSpy) OnDelete(obj interface{}) { es.deletes = append(es.deletes, obj) }
+
+func TestExecPluginViaInformer(t *testing.T) {
+	result, clientAuthorizedToken, clientCertFileName, clientKeyFileName := startTestServer(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*60)
+	defer cancel()
+
+	adminClient := clientset.NewForConfigOrDie(result.ClientConfig)
+	ns := createNamespace(ctx, t, adminClient)
+
+	tests := []struct {
+		name                          string
+		clientConfigFunc              func(*rest.Config)
+		wantAuthorizationHeaderValues [][]string
+		wantCertificate               *tls.Certificate
+	}{
+		{
+			name: "authorized token",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = []clientcmdapi.ExecEnvVar{
+					{
+						Name: outputEnvVar,
+						Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"token": %q
+							}
+						}`, clientAuthorizedToken),
+					},
+				}
+			},
+		},
+		{
+			name: "authorized certificate",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = []clientcmdapi.ExecEnvVar{
+					{
+						Name: outputEnvVar,
+						Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"clientCertificateData": %s,
+								"clientKeyData": %s
+							}
+						}`, read(t, clientCertFileName), read(t, clientKeyFileName)),
+					},
+				}
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			clientConfig := rest.AnonymousClientConfig(result.ClientConfig)
+			clientConfig.ExecProvider = &clientcmdapi.ExecConfig{
+				Command: "testdata/exec-plugin.sh",
+				// TODO(ankeesler): move to v1 once exec plugins go GA.
+				APIVersion: "client.authentication.k8s.io/v1beta1",
+			}
+
+			if test.clientConfigFunc != nil {
+				test.clientConfigFunc(clientConfig)
+			}
+			client := clientset.NewForConfigOrDie(clientConfig)
+
+			informerSpy := startConfigMapInformer(ctx, t, client, ns.Name)
+			createUpdateDeleteConfigMap(ctx, t, client.CoreV1().ConfigMaps(ns.Name))
+
+			// Validate that the informer was called correctly.
+			if want, got := 1, len(informerSpy.adds); want != got {
+				t.Errorf("wanted %d add event(s), got %d (%s)", want, got, informerSpy.adds)
+			}
+			if want, got := 1, len(informerSpy.updates); want != got {
+				t.Errorf("wanted %d update event(s), got %d (%s)", want, got, informerSpy.updates)
+			}
+			if want, got := 1, len(informerSpy.deletes); want != got {
+				t.Errorf("wanted %d add event(s), got %d (%s)", want, got, informerSpy.deletes)
+			}
```

Comment:
Assert that the events match what you expect to see.

Reply:
e2105d431ee